experimental/relativetime/src/format.rs

```diff
@@ -0,0 +1,85 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use std::fmt::Write;
+
+use fixed_decimal::{FixedDecimal, Sign};
+use writeable::Writeable;
+
+use crate::{
+    options::{Numeric, RelativeTimeFormatterOptions},
+    relativetime::RelativeTimeFormatter,
+};
+
+pub mod parts {
+    use writeable::Part;
+
+    /// The [`Part`] used by [`FormattedRelativeTime`](crate::FormattedRelativeTime) to mark the
+    /// part of the string that is without a placeholder.
+    pub const LITERAL: Part = Part {
+        category: "relativetime",
+        value: "literal",
+    };
+}
+
+/// An intermediate structure returned by [`RelativeTimeFormatter`](crate::RelativeTimeFormatter).
+/// This structure can be consumed via [`Writeable`](Writeable) trait to a string or buffer.
+pub struct FormattedRelativeTime<'a> {
+    pub(crate) formatter: &'a RelativeTimeFormatter,
+    pub(crate) options: &'a RelativeTimeFormatterOptions,
+    pub(crate) value: FixedDecimal,
+}
+
+impl<'a> Writeable for FormattedRelativeTime<'a> {
+    fn write_to_parts<S: writeable::PartsWrite + ?Sized>(&self, sink: &mut S) -> core::fmt::Result {
+        if self.options.numeric == Numeric::Auto {
+            let _ = &self.formatter.rt.get().relatives;
+        }
+
+        let plural_rules_mapping = if self.value.sign() == Sign::Negative
+            || (self.value.sign() == Sign::Negative && self.value.is_zero())
+        {
+            &self.formatter.rt.get().past
+        } else {
+            &self.formatter.rt.get().future
+        };
+        let category = self.formatter.plural_rules.category_for(&self.value);
+        let pattern = match category {
+            icu_plurals::PluralCategory::Zero => &plural_rules_mapping.zero,
+            icu_plurals::PluralCategory::One => &plural_rules_mapping.one,
+            icu_plurals::PluralCategory::Two => &plural_rules_mapping.two,
+            icu_plurals::PluralCategory::Few => &plural_rules_mapping.few,
+            icu_plurals::PluralCategory::Many => &plural_rules_mapping.many,
+            icu_plurals::PluralCategory::Other => &plural_rules_mapping.other,
+        };
+
+        // Default to using PluralCategory::Other mapping.
+        let pattern = pattern.as_ref().unwrap_or_else(|| {
+            plural_rules_mapping
+                .other
+                .as_ref()
+                .expect("Mapping for PluralCategory::Other must be present.")
```

Comment:
Issue: `.expect` is too strong here. This is a good example of something you should check at deserialization time. See https://github.com/unicode-org/icu4x/pull/2768

Reply:
Making `other` a required field in [relativetime/src/provider.rs](https://github.com/unicode-org/icu4x/blob/main/experimental/relativetime/src/provider.rs#L93)